### PR TITLE
*: specialize fmt::formatter<> for formatting using {fmt}

### PR DESCRIPTION
**Pull request by tchaikov**
in {fmt} v9, fmt::formatter<> is not defined automatically for the types with operator<<, we need to define them explictly. so in this change, the types printed using `fmt::print()` or `fmt::format()` are equiped with fmt::formatter<> to appease {fmt} v9.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
the change on crimson side is created separately at #49000 in hope that these two PRs can be reviewed and merged more quickly

**Comment by tchaikov** [comment]
jenkins test make check

File: src/msg/Message.h [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #ifndef CEPH_MESSAGE_H
 #define CEPH_MESSAGE_H
 
+#include <concepts>
```
Comment (ronen-fr) [inline src/msg/Message.h:18]: do we have the include file in all our target compilers?

File: src/msg/Message.h [hunk 2]
```diff
@@ -570,4 +571,9 @@ MURef<T> make_message(Args&&... args) {
   return {new T(std::forward<Args>(args)...), TOPNSPC::common::UniquePtrDeleter{}};
 }
 }
+
+#if FMT_VERSION >= 90000
+template <std::derived_from<Message> M> struct fmt::formatter<M> : fmt::ostream_formatter {};
```
Comment (ronen-fr) [inline src/msg/Message.h:576]: nice!

File: src/common/ceph_time.h [hunk 3]
```diff
@@ -533,4 +533,9 @@ template<typename Rep, typename Period>
 ostream& operator<<(ostream& m, const chrono::duration<Rep, Period>& t);
 }
 
+#if FMT_VERSION >= 90000
```
Comment (ronen-fr) [inline src/common/ceph_time.h:536]: That's the only modified file I wasn't sure about:
I'd rather use the fmtlib in reverse: use fmtlib for operator<<

**Review by ronen-fr — APPROVED** [review]
LGTM apart from some questions.
@tchaikov - your call whether to change.

And re the utime.h comment: note the existence of utime_fmt.h, and the
special formatting specifiers used there

File: src/common/ceph_time.h (hunk 3 above)
Comment (tchaikov) [reply src/common/ceph_time.h:536]: i don't feel strong either way. but i'd like to focus on enabling ceph to build with fmt v9 first. and then try to reimplement the formatter.

File: src/msg/Message.h (hunk 1 above)
Comment (tchaikov) [reply src/msg/Message.h:18]: `concepts` should be provided by a standard library compliant to C++20. and we already use it in, for instance, https://github.com/ceph/ceph/blob/021fa45573caa22c95fc3f8173b21f255b3d93b2/src/include/intarith.h#L20 .

**Comment by tchaikov** [comment]
@ronen-fr hi Ronen, thank you for your review. i just force-pushed the branch with a couple more `fmt::formatter` changes -- they were missing in the revision you reviewed.

**Comment by tchaikov** [comment]
> LGTM apart from some questions. @tchaikov - your call whether to change.
> 
> And re the utime.h comment: note the existence of utime_fmt.h, and the special formatting specifiers used there

@ronen-fr thanks for pointing me to this file. it will serve as a good reference once we need to reimplement the formatters for `chrono` types.

**Comment by tchaikov** [comment]
being built at https://shaman.ceph.com/builds/ceph/wip-crimson-fmt-v9/69195f1b698bb6460018ee738f5cf5fe2bc1c093/

**Pull request merged**